cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillController.java

```diff
@@ -115,23 +116,27 @@ public void release() {
     executionService.stop();
   }
 
-  public void complete() {
+  public void complete () {
     terminateOnServiceStop = true;
     executionService.stop();
-
     try {
+      RuntimeJobStatus status;
       RetryStrategy retryStrategy = RetryStrategies.timeLimit(
         5, TimeUnit.SECONDS, RetryStrategies.exponentialDelay(500, 2000, TimeUnit.MILLISECONDS));
 
       // Make sure the remote execution is completed
       // Give 5 seconds for the remote process to shutdown. After 5 seconds, issues a kill.
       long startTime = System.currentTimeMillis();
-      while (Retries.callWithRetries(remoteProcessController::isRunning, retryStrategy, Exception.class::isInstance)) {
+      while ((status = Retries.callWithRetries(
+                remoteProcessController::getStatus, retryStrategy, Exception.class::isInstance))
+                                        == RuntimeJobStatus.RUNNING) {
         if (System.currentTimeMillis() - startTime >= 5000) {
           throw new IllegalStateException("Remote process for " + programRunId + " is still running");
         }
         TimeUnit.SECONDS.sleep(1);
       }
+      LOG.debug("Force termination of remote process for program run {} with status {}", programRunId, status);
+      remoteProcessController.kill(status);
```

Comment:
why remoteProcessController.kill(status) instead of status?

Reply:
We will be using the status to figure out if the program should be killed or not. 
We do getStatus in the while loop. In case of tethering, we always get UNKNOWN. So we will use it to not issue a kill() on the control channel.